x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RequestXContent.java

```diff
@@ -114,82 +126,121 @@ private static ObjectParser<EsqlQueryRequest, Void> objectParserAsync(Supplier<E
         return parser;
     }
 
-    private static List<TypedParamValue> parseParams(XContentParser p) throws IOException {
-        List<TypedParamValue> result = new ArrayList<>();
+    private static QueryParams parseParams(XContentParser p) throws IOException {
+        List<QueryParam> result = new ArrayList<>();
+        List<String> errors = new ArrayList<>();
         XContentParser.Token token = p.currentToken();
+        boolean namedParameter = false;
+        boolean unnamedParameter = false;
 
         if (token == XContentParser.Token.START_ARRAY) {
             Object value = null;
-            String type = null;
-            TypedParamValue previousParam = null;
-            TypedParamValue currentParam;
+            DataType type = null;
+            QueryParam currentParam = null;
+            TempObjects param;
 
             while ((token = p.nextToken()) != XContentParser.Token.END_ARRAY) {
                 XContentLocation loc = p.getTokenLocation();
 
                 if (token == XContentParser.Token.START_OBJECT) {
                     // we are at the start of a value/type pair... hopefully
-                    currentParam = PARAM_PARSER.apply(p, null);
+                    param = PARAM_PARSER.apply(p, null);
+                    if (param.fields.size() > 1) {
+                        errors.add("Multiple parameters in one curly bracket is not supported.");
+                    }
+                    for (Map.Entry<String, Object> entry : param.fields.entrySet()) {
+                        if (isValidParamName(entry.getKey()) == false) {
+                            errors.add(
+                                entry.getKey()
+                                    + " is not a valid parameter name. "
+                                    + "A valid parameter name starts with a letter and contains letters, digits and underscores only."
+                            );
+                        }
+                        type = EsqlDataTypes.fromJava(entry.getValue());
+                        if (type == null) {
+                            errors.add(loc + " " + entry + " is not supported as a parameter.");
```

Comment:
`loc`? This results in error messages like `"reason": "Failed to parse params: [3:39 z1=[1, x, 3] is not supported as a parameter.]",`

Also, food for thought: the location of the faulty elements while parsing the params.... can we do better here by incorporating it? In this case it seems like an inclusion by mistake, but will it help if we add the location to every error message from parsing the parameters?

Reply:
I'll add loc to each error message from parsing the parameters.